### PR TITLE
Creation of Connection/Channel objects

### LINKED ISSUE #42: Create connection objects
Connections at the moment are booleans (connected or not). We should create an object, one for classical one for quantum, that models the connections better.

### PR DESCRIPTION
**Pull request by AnirbanHFX**
This branch adds quantum and classical connection objects to QuNetSim. 

components/host.py has been modified to add these objects instead of only receiver_ids when new connections are added.

These objects are flexible and a user can change the connection model as required. Currently the quantum connection defaults to the fiber model, and the classical connection defaults to a model which only contains length and transmission probability as parameters.

Here is an example of how a model is accessed, changed or modified -
```
host1.add_connection(host2.host_id)                                         #Initialize the connection
host1.quantum_connections[host2.host_id].model = Fibre_Model()              #Set Fibre_Model as the model (already set by default)
host1.quantum_connections[host2.host_id].model.length = 10                  #Set length (0 by default)
host1.quantum_connections[host2.host_id].model.alpha = 0.5                  #Set alpha (0 by default)
print(host1.quantum_connections[host2.host_id].model.transmission_p)        #Print transmission probability (1 by default)
```

Note : Transmission probability checking and dropping of qubits/packets have not been added yet. These changes have to be made to components/network.py and will be done later.

### CODE REVIEW

**Comment by AnirbanHFX** [comment]
Added a simple model of Binary Erasure Channel to qunetsim/objects/quantum_connections.py.

Here is an example of how to switch to the BEC model -
```
from qunetsim.objects import Binary_Erasure_Model
.
.
.
host1.add_connection(host2.host_id)
host1.quantum_connections[host2.host_id].model = Binary_Erasure_Model(probability = x)
```

Also added checking of channel properties in the _process_queue of network.py. For example, if a qubit is erased by the BEC, then the packet payload becomes 'None'. For this, the _process_queue checks the return value of a function titled 'qubit_func' defined in the channel model.

Here is the code snippet used for this checking -
```
# Simulate Channel Properties
if packet.payload_type == Constants.QUANTUM:
     # Check if direct connection exists
     if receiver in sending_host.quantum_connections:
         # Modify the packet according to the qubit_func method described by the model
         packet.payload = sending_host.quantum_connections[receiver].model.qubit_func(packet.payload)
         # Abort the loop if qubit has been lost
         if packet.payload is None:
               continue
```

Any channel model must contain a method titled qubit_func that takes, as input, a qubit object and returns a modified qubit object or None.

**Comment by stephendiadamo** [comment]
I'll start reviewing this pull request this week

**Comment by stephendiadamo** [comment]
@HalcyonicFalconX While I'm reviewing the code, it would be helpful to already start adding test cases for your features. As a minimum requirement, each added feature will need a unit test. You can see our current tests in the `integration_tests` folder.

**Comment by AnirbanHFX** [comment]
Added integration tests for the channel object.

File: integration_tests/test_channel.py
Comment (benjione) [inline integration_tests/test_channel.py]: Comments can be deleted

File: qunetsim/objects/classical_connection.py
Comment (benjione) [inline qunetsim/objects/classical_connection.py]: inherit from object, to be more consistent with the other code.

File: qunetsim/objects/quantum_connection.py
Comment (benjione) [inline qunetsim/objects/quantum_connection.py]: inherit from object, to be more consistent with the other code.

File: qunetsim/objects/quantum_connection.py
Comment (benjione) [inline qunetsim/objects/quantum_connection.py]: Maybe add an optional model parameter here, which defaults to Fibre_Model ?

File: qunetsim/components/network.py
Comment (benjione) [inline qunetsim/components/network.py]: Look into _route_quantum_info, there we currently have error parameters. These have to be removed from the network. I also would put the channel part into this function, not here.

**Review by benjione — CHANGES_REQUESTED** [review]
Please clean up the network.py. You have to delete the error parameters there and put the channel model into the route_quantum_info method.

File: qunetsim/objects/classical_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/classical_connection.py]: Done.

File: qunetsim/objects/quantum_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/quantum_connection.py]: Done.

File: qunetsim/objects/quantum_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/quantum_connection.py]: Added an optional parameter.

File: integration_tests/test_channel.py
Comment (AnirbanHFX) [reply integration_tests/test_channel.py]: Done.

File: qunetsim/components/network.py
Comment (AnirbanHFX) [reply qunetsim/components/network.py]: Error parameters have been removed.
Application of channel properties has been moved to transfer_qubits function in the _route_quantum_info method.

I would also like to point out that though _route_quantum_info takes a list of qubits as a parameter, it is never called with a list containing more than 1 qubit. Can you clarify as to why a list is required?

**Review by benjione — COMMENTED** [review]
Looks good to me now. Since these are bigger changes, let's wait for @stephendiadamo to also review it.

**Review by benjione — APPROVED** [review]

File: qunetsim/components/network.py
Comment (stephendiadamo) [reply qunetsim/components/network.py]: @HalcyonicFalconX  This is kind of "future proofing" thing I thought about a while ago. I thought there might come a time when we want to add a feature for "sending a list of qubits" at some point, therefore the network takes a list of qubits instead of 1 qubit.

**Review by stephendiadamo — CHANGES_REQUESTED** [review]
@HalcyonicFalconX 
I've checked out the forked version of the code and made some changes to it, but can't push since I'm not a collaborator on that fork. 

Could you add me to it? There are some changes that I have in mind, but we can build on the changes I already made.

**Comment by AnirbanHFX** [comment]
> @HalcyonicFalconX
> I've checked out the forked version of the code and made some changes to it, but can't push since I'm not a collaborator on that fork.
> 
> Could you add me to it? There are some changes that I have in mind, but we can build on the changes I already made.

I have enabled the "Allow edits and access to secrets by maintainers" option, so you should now be able to commit to the channel branch of my fork.

File: qunetsim/objects/classical_connection.py
Comment (stephendiadamo) [inline qunetsim/objects/classical_connection.py]: Probably meters is more practical than kilometers in this case

File: qunetsim/components/network.py
Comment (stephendiadamo) [inline qunetsim/components/network.py]: `if not` instead of `is False`.

File: qunetsim/components/network.py
Comment (stephendiadamo) [inline qunetsim/components/network.py]: `if not` instead of `is False`

File: qunetsim/objects/quantum_connection.py
Comment (stephendiadamo) [inline qunetsim/objects/quantum_connection.py]: Default should be the identity channel

**Review by stephendiadamo — CHANGES_REQUESTED** [review]
I think everything is in order now and we can start to think about some changes to channel objects.

- Since there are two channel types (i.e. classical and quantum) that will have very similar properties, there should be a `channel` object in which both the classical and quantum channel extends
- For specific channel models, we should create a separate folder nested within the `objects` folder and create a file for each specific model instead of putting the models into the channel objects 
- The channels should have a sender and a receiver ID. The channel shouldn't be owned by any single sender or receiver and these input/outputs should be able to change at some point. For now, adding a sender property to the channels is enough

Let's leave it there for now. I think we can merge it in after these changes.

File: qunetsim/objects/quantum_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/quantum_connection.py]: BEC with erasure probability = 0 set as default.

File: qunetsim/objects/classical_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/classical_connection.py]: Changed Km to m.

File: qunetsim/objects/connections/classical_connection.py
Comment (stephendiadamo) [inline qunetsim/objects/connections/classical_connection.py]: Here should extend the `connection` object i.e
```
class ClassicalConnection(Connection)
```
See: https://www.python-course.eu/python3_inheritance.php

File: qunetsim/objects/connections/quantum_connection.py
Comment (stephendiadamo) [inline qunetsim/objects/connections/quantum_connection.py]: Similarly here, should inherit from `connection`

Comment (stephendiadamo) [reply qunetsim/objects/connections/quantum_connection.py]: Then you can remove the setters and getters that overlap between the two objects since they'll be inherited

**Review by stephendiadamo — CHANGES_REQUESTED** [review]
The two connection objects should extend the `connection` object. Inherited methods that are not modified should be deleted.

File: qunetsim/objects/connections/classical_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/connections/classical_connection.py]: Done.

File: qunetsim/objects/connections/quantum_connection.py
Comment (AnirbanHFX) [reply qunetsim/objects/connections/quantum_connection.py]: Done

**Review by stephendiadamo — APPROVED** [review]

**Pull request merged**
